java/flight/src/main/java/org/apache/arrow/flight/example/integration/IntegrationTestClient.java

```diff
@@ -94,11 +96,34 @@ private void run(String[] args) throws ParseException, IOException {
       jsonRoot = VectorSchemaRoot.create(root.getSchema(), allocator);
       VectorUnloader unloader = new VectorUnloader(root);
       VectorLoader jsonLoader = new VectorLoader(jsonRoot);
-      FlightClient.ClientStreamListener stream = client.startPut(descriptor, root);
+      FlightClient.ClientStreamListener stream = client.startPut(descriptor, root, reader,
+          new StreamListener<PutResult>() {
+            int counter = 0;
+
+            @Override
+            public void onNext(PutResult val) {
+              final String metadata = StandardCharsets.UTF_8.decode(val.getApplicationMetadata()).toString();
+              if (!Integer.toString(counter).equals(metadata)) {
+                throw new RuntimeException(
+                    String.format("Invalid ACK from server. Expected '%d' but got '%s'.", counter, metadata));
+              }
+              counter++;
+            }
+
+            @Override
+            public void onError(Throwable t) {
+            }
+
+            @Override
+            public void onCompleted() {
```

Comment:
Shouldn't this test wait until onCompleted is called for all messages (to avoid flaky test issues)?